### PR TITLE
Add channel binding support for SASL + TLS

### PR DESCRIPTION
**Pull request by krizex**
When building the connection base on SASL(Kerberos) + TLS to connect a Windows
ldap server (aka Domain Controller), we have to enforce ldap channel binding for
the connection otherwise the server will reject the connection.
See: https://support.microsoft.com/en-us/help/4520412/2020-ldap-channel-binding-and-ldap-signing-requirements-for-windows

This change also consider the useage of the SASL without TLS connection, so
won't calculate the channel binding token for that case.

Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

**Comment by krizex** [comment]
This is a fix for https://github.com/cannatag/ldap3/issues/831

**Comment by krizex** [comment]
I have verified this fix against my Windows Server 2012R2

**Comment by cannatag** [comment]
Thanks, will be included in the next version of ldap3.

Bye,
Giovanni

**Comment by krizex** [comment]
Thank you @cannatag , may I know the date for the next release? I'd like to integrate this feature into my product.

**Comment by krizex** [comment]
Hi @cannatag , Can you take a look at this PR when you get time? Thanks in advance!

File: ldap3/protocol/sasl/kerberos.py [hunk 1]
```diff
@@ -26,22 +26,54 @@
...
+        from cryptography import x509
+        from cryptography.hazmat.backends import default_backend
+        from cryptography.hazmat.primitives import hashes
+        from cryptography.exceptions import UnsupportedAlgorithm
+    except ImportError:
+        raise LDAPPackageUnavailableError('package cryptography missing')
+    cert = x509.load_der_x509_certificate(server_certificate, default_backend())
+    hash_algorithm = cert.signature_hash_algorithm
+    if hash_algorithm.name in ('md5', 'sha1'):
+        digest = hashes.Hash(hashes.SHA256(), default_backend())
```
Comment (BobBall) [inline ldap3/protocol/sasl/kerberos.py:69]: Reference rfc5929?

File: ldap3/protocol/sasl/kerberos.py [hunk 2]
```diff
@@ -26,22 +26,54 @@
...
 
 
+def get_channel_bindings(ssl_socket):
+    try:
+        server_certificate = ssl_socket.getpeercert(True)
+    except:
+        # it is not SSL socket
+        return None
+    try:
+        from cryptography import x509
```
[the remark this answers is not in the document]
Comment (krizex) [reply ldap3/protocol/sasl/kerberos.py:60]: Will remove the lines of 43~46

File: ldap3/protocol/sasl/kerberos.py [hunk 3]
```diff
@@ -26,22 +26,54 @@
 # original code by Hugh Cole-Baker, modified by Peter Foley
 # it needs the gssapi package
 import socket
+import hashlib
+import base64
```
[the remark this answers is not in the document]
Comment (krizex) [reply ldap3/protocol/sasl/kerberos.py:30]: Yes, will remove.

File: ldap3/protocol/sasl/kerberos.py (hunk 1 above)
Comment (krizex) [reply ldap3/protocol/sasl/kerberos.py:69]: Yes, will add inline comments

File: ldap3/protocol/sasl/kerberos.py [hunk 4]
```diff
@@ -26,22 +26,54 @@
...
+    try:
+        server_certificate = ssl_socket.getpeercert(True)
+    except:
+        # it is not SSL socket
+        return None
+    try:
+        from cryptography import x509
+        from cryptography.hazmat.backends import default_backend
+        from cryptography.hazmat.primitives import hashes
+        from cryptography.exceptions import UnsupportedAlgorithm
```
[the remark this answers is not in the document]
Comment (krizex) [reply ldap3/protocol/sasl/kerberos.py:63]: Will remove.

**Comment by cannatag** [comment]
Please, can you send the latest merge request to the "dev" branch?

**Comment by krizex** [comment]
@cannatag OK, close this one.

**Pull request closed without merging**